crates/proto/src/serialize/binary/decoder.rs

```diff
@@ -157,25 +155,21 @@ impl<'a> BinDecoder<'a> {
     ///
     /// The slice of the specified length, otherwise an error
     pub fn read_slice(&mut self, len: usize) -> ProtoResult<Restrict<&'a [u8]>> {
-        let end = self
-            .index
-            .checked_add(len)
-            .ok_or_else(|| ProtoError::from("invalid length for slice"))?;
-        if end > self.buffer.len() {
+        if len > self.remaining.len() {
             return Err("buffer exhausted".into());
         }
-        let slice: &'a [u8] = &self.buffer[self.index..end];
-        self.index += len;
-        Ok(Restrict::new(slice))
+        let (read, remaining) = self.remaining.split_at(len);
+        self.remaining = remaining;
+        Ok(Restrict::new(read))
     }
 
     /// Reads a slice from a previous index to the current
     pub fn slice_from(&self, index: usize) -> ProtoResult<&'a [u8]> {
-        if index > self.index {
+        if index > self.index() {
```

Comment:
Here, too, I'd prefer a more direct expression of what we're checking for here. Also I know this is already there, but maybe rewrite that error message? I'm not a native speaker, but I'm pretty sure "antecedes" is fairly obscure.